stream: allow typed arrays to be written and read

SirR4T opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Partially fixing #1826 

## Caveat
- Any `TypedArrays` can be written, now, but in non-`objectMode`, the `.write()` method still receives raw `Buffer`s.

I'm not completely sure that is the behaviour we want, but I'm willing to work towards fixing it, with some mentoring.

SirR4T commented:
on second thought, the `Uint8Array` support for streams does the same thing as well, so 🤷‍♂️

mscdex commented on lib/stream.js:
```diff
@@ -45,37 +45,50 @@ Stream.Stream = Stream;
+      '[object Float32Array]',
+      '[object Float64Array]',
+      '[object DataView]'
+    ].includes(Object.prototype.toString.call(obj));
```
This is going to be pretty slow. Surely there is a better way...

SirR4T replied:
Will `.indexOf()` be faster? Open to other suggestions as well.

mscdex replied:
Can't we just use something like `ArrayBuffer.isView()`? It seems to be available since at least node v4. The reason the string check was used before was because we were explicitly checking for only `Uint8Array`.

BridgeAR replied:
On `util.types` are a lot of checks that can be used as well.

mscdex replied:
@BridgeAR the relevant function from `util.types` is already being used earlier in this file if it exists. This implementation here is used if either it or the function from the `util` binding are not available.

SirR4T replied:
@mscdex so `ArrayBuffer.isView()` could still be available, even though `util.types` and `util` binding are not available?
In other PRs dealing with typed arrays, i have been using:
```
const { isArrayBufferView } = require('internal/util/types');
```
Same can be used here too?
Wasn't sure whether compatibility with `0.x` versions was specifically required here, and how to add that support.

SirR4T replied:
So i'm guessing current PR stays put?

Or is there any merit in adding a patch of this sort:
```
diff --git a/lib/stream.js b/lib/stream.js
index c6861b3464..58aab7c16d 100644
--- a/lib/stream.js
+++ b/lib/stream.js
@@ -42,6 +42,15 @@ Stream.finished = eos;
 Stream.Stream = Stream;

 // Internal utilities
+try {
+  const { isArrayBufferView } = require('internal/util/types');
+  if (typeof isArrayBufferView === 'function') {
+    Stream._isArrayBufferView = isArrayBufferView;
+  }
+} catch (e) {
+}
+
+if (!Stream._isArrayBufferView) {
   try {
     const types = require('util').types;
     if (types && typeof types.isArrayBufferView === 'function') {
@@ -53,6 +62,7 @@ try {
     }
   } catch (e) {
   }
+}

 if (!Stream._isArrayBufferView) {
   Stream._isArrayBufferView = function _isArrayBufferView(obj) {
```

mscdex replied:
@SirR4T That's more or less the exact same thing that's being done below that with `require('util').types`. I still think using `ArrayBuffer.isView()` as the fallback is probably best. I suppose the function could be cached at startup like `internal/util/types` does.

SirR4T replied:
Updated to use `ArrayBuffer.isView`. Although do verify, if simply having it in a constant will cache it? Or should the `const isTypedArray` be moved to module scope?

Also, shouldn't we be checking for existence of `ArrayBuffer.isView` first, and then be falling back to other methods (`util`.types)?

mscdex commented on lib/stream.js:
```diff
@@ -45,37 +45,50 @@ Stream.Stream = Stream;
   const types = require('util').types;
   if (types && typeof types.isUint8Array === 'function') {
-    Stream._isUint8Array = types.isUint8Array;
+    // Stream._isUint8Array = types.isUint8Array;
```
These comments should be removed

SirR4T replied:
done.

BridgeAR commented on lib/stream.js:
```diff
@@ -45,37 +45,50 @@ Stream.Stream = Stream;
-    Stream._uint8ArrayToBuffer = function _uint8ArrayToBuffer(chunk) {
+  if (!Stream._typedArrayToBuffer) {
+    Stream._typedArrayToBuffer = function _typedArrayToBuffer(chunk) {
       return Buffer.prototype.slice.call(chunk);
```
This will likely not work with anything besides `Uint8Array`. So any other TypedArray would fail.

SirR4T replied:
updated `._typedArrayToBuffer` to use `Buffer.from()`.

mscdex commented on lib/stream.js:
```diff
@@ -45,37 +45,50 @@ Stream.Stream = Stream;
 try {
   const types = require('util').types;
   if (types && typeof types.isUint8Array === 'function') {
```
This needs to be updated to check for the correct function.

SirR4T replied:
done.

SirR4T commented on lib/stream.js:
```diff
@@ -87,7 +87,7 @@ if (version[0] === 0 && version[1] < 12) {
   if (!Stream._typedArrayToBuffer) {
     Stream._typedArrayToBuffer = function _typedArrayToBuffer(chunk) {
-      return Buffer.prototype.slice.call(chunk);
+      return Buffer.from(chunk.buffer, chunk.byteOffset, chunk.byteLength);
```
the `Buffer.prototype.slice()` would have made a different Buffer. With `Buffer.from()`, the underlying memory will still be shared. Will that be a problem?

mscdex commented on lib/stream.js:
```diff
@@ -44,39 +44,39 @@ Stream.Stream = Stream;
+if (!Stream._isArrayBufferView) {
+  // Cached to make sure no userland code can tamper with it.
+  const isArrayBufferView = ArrayBuffer.isView;
+  Stream._isArrayBufferView = isArrayBufferView;
```
Oh, in this case we don't actually need to do any caching. I think we can just do `Stream._isArrayBufferView = isArrayBufferView = ArrayBuffer.isView;`

SirR4T replied:
done.

SirR4T commented:
@mscdex / @BridgeAR : any other changes suggested / required here?

mscdex commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16777/

SirR4T commented:
Hi @mscdex , sorry for the commit and CI noise, but I had rebased this branch against master, after [the earlier CI run](https://ci.nodejs.org/job/node-test-pull-request/16777/) had completed.

For this to be merged in, should another CI run be initiated?

Also, kind of off topic: 
I personally prefer to keep my feature branches updated with (rebased against) master, so that builds across different feature branches may benefit simultaneously. Is this practice frowned upon? Or are there other ways to achieve the same?

mcollina commented on lib/stream.js:
```diff
@@ -44,39 +44,37 @@ Stream.Stream = Stream;
+    Stream._isArrayBufferView = types.isArrayBufferView;
   } else {
     // This throws for Node < 4.2.0 because there's no util binding and
     // returns undefined for Node < 7.4.0.
```
This comment would need to be updated. When has `isArrayBufferView`  been added?

mscdex replied:
v7.4.0.

mcollina commented on doc/api/stream.md:
```diff
@@ -1860,15 +1872,19 @@ It can be overridden by child classes but it **must not** be called directly.
+* `chunk` {Buffer|TypedArray|DataView|string|null|any} Chunk of data to push
+  into the read queue. For streams not operating in object mode, `chunk` must
+  be a string, `Buffer` or `Uint8Array`. For object mode streams, `chunk` may
+  be any JavaScript value.
```
I guess this sentence is not up-to-date.  It's not just string, `Buffer` or `Uint8Array` anymore.

SirR4T replied:
True, but the typed arrays and data views can be passed only in object mode. So thought that the current statement still remains true.

mcollina replied:
Maybe I misunderstood this PR then.  I thought the whole point of this is to support all `TypedArray` and `DataView`  in binary mode. 

This looks like from looking at the code  as `!state.objectMode && Stream._isArrayBufferView` works when `objectMode` is `false.

SirR4T replied:
yea... I wasn't sure of the behaviour we required here, which is why I had raised the PR [with a caveat](https://github.com/nodejs/node/pull/22427#issue-209744259).

[The PR](https://github.com/nodejs/node/pull/11608) introducing support for `Uint8Array` in streams, too, [asserts a similar behaviour](https://github.com/nodejs/node/blob/master/test/parallel/test-stream-uint8array.js#L11-L47), so I wasn't sure which behaviour was required here.

mcollina replied:
Exactly. Still, this PR make accepting `TypedArray` and `DataView`  everywhere so it should be reflected in the docs.

mcollina commented on lib/stream.js:
```diff
@@ -44,39 +44,37 @@ Stream.Stream = Stream;
+      return Buffer.from(chunk.buffer, chunk.byteOffset, chunk.byteLength);
     };
   }
 }
```
I think we can clean up this whole file. As long as it exports the two functions we need, it's ok to not have all those checks. readable-stream@3 supports Node 6+ anyway.

mcollina commented:
I've added the dont-land-on-6 and dont-land-on-8 labels out of caution.

mcollina commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/16982/

addaleax commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/16988/

mcollina commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/17003/
CITGM (master): https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1527/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/1​52​8

mcollina commented on test/parallel/test-stream-typedarrays.js:
```diff
@@ -0,0 +1,112 @@
+
+  const out = readable.read();
+  assert.strictEqual(out, 'ABCDEFGH'.repeat(10));
+}
```
Can you please add a test where a Readable and a Writable are piped together? Also, a test that includes a Transform?

SirR4T commented:
- [x] removed legacy compatibility code for lib/stream.js (cc: @mcollina , @mscdex , and @addaleax due to comments [here](https://github.com/nodejs/node/pull/11608/files#r103704734) and [here](https://github.com/nodejs/node/pull/22427#discussion_r214599788))
- [x] added tests for streams where writeable is piped to readable, and with transforms,
- [x] updated docs for typed arrays and data views,
- [x] added docs explicitly stating conversion of typed arrays and data views in non-object mode.
- [x] rebased against current master,

SirR4T commented:
@mscdex : pushed changes with `objectMode: false`, and some changes to internal `_stream_readable.js` and `_stream_writable.js` to remove checks which I think are unnecessary now. They might have been necessary before `Buffer` was a subclass of `Uint8Array`.

One test is still failing, which raises this question: how should `encoding` work, for any `.write()` methods where chunk is not a string?

mcollina commented:
> pushed changes with objectMode: false, and some changes to internal _stream_readable.js and _stream_writable.js to remove checks which I think are unnecessary now. They might have been necessary before Buffer was a subclass of Uint8Array.

Are you sure? Can you please add some checks in the tests that everything that is emitted/returned by `on('data')` and `read()` are `Buffer`s? Also, can you check that everything that gets into `_write()` and `_writev()` are indeed `Buffer`?

> One test is still failing, which raises this question: how should encoding work, for any .write() methods where chunk is not a string? As far as I understand, it works with Buffers and Uint8Arrays currently (before this PR), but will not work with any other typed array or data view.

I think this is why you need the conversion that you have commented out.

lundibundi commented:
ping @SirR4T.

SirR4T commented:
Currently swamped at work, will look into this whenever I find some time.
Most likely within the next two weeks.
Sarat.



On Wed, Oct 3, 2018 at 2:01 PM Denys Otrishko <notifications@github.com>
wrote:

> ping @SirR4T <https://github.com/SirR4T>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/nodejs/node/pull/22427#issuecomment-426553419>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB8IMLS1rZyf_9rPyjgl5jI0WFR1C_jbks5uhHXVgaJpZM4WFUJ6>
> .
>

antsmartian commented:
@SirR4T Ping, any update on this?

SirR4T commented:
fixed the merge conflicts for now, but if someone could guide me on correcting the test cases, would appreciate the help.

oyyd commented on test/parallel/test-stream-typedarrays.js:
```diff
@@ -0,0 +1,113 @@
+    write: common.mustCall((chunk, encoding, cb) => {
+      assert(!(chunk instanceof Buffer));
+      assert(ArrayBuffer.isView(chunk));
+      assert.deepStrictEqual(chunk, nthWrittenBuffer(n++));
```
Instances of `Buffer` could not be `deepStrictEqual ` to those of `TypedArray`, could they?

Maybe try:
```js
assert.deepStrictEqual(Buffer.from(chunk.buffer, chunk.byteOffset, chunk.length), nthWrittenBuffer(n++));
```

BridgeAR commented:
@SirR4T what is the status here?

jasnell commented:
Given that this is out of date and there has been no activity in over a year, closing. Can reopen if it is picked back up again
